WIP: Add libradolan recipe

kmuehlbauer opened the pull request:
This adds `libradolan` recipe for linux and osx. Windows can be added, if this is converted to a feedstock.

This is WIP yet, I'll ping if this can be reviewed (or if I need help, which will likely be the case).

conda-forge-linter commented:
Hi! This is the friendly automated conda-forge-linting service.

I just wanted to let you know that I linted all conda-recipes in your PR (```recipes/libradolan```) and found it was in an excellent condition.

kmuehlbauer commented:
ping @JuergenSimon  I've added you as an maintainer. If this is converted into feedstock you will get a notification to join conda-forge.

kmuehlbauer commented:
OK, we have two problems:

1. Linux Build: boost is complaining a missing `quadmath.h`.  
2. OSX is having issues with the architecture: `ld: symbol(s) not found for architecture x86_64`

ping @conda-forge/boost It seems that I'm missing some package, but which one? (It's imho not a boost issue, but you might point me into the right direction)

@ocefpaf Do you have any ideas, on how to address the architecture issue?

ocefpaf commented on recipes/libradolan/build.sh:
```diff
@@ -0,0 +1,8 @@
+#!/usr/bin/env bash
+
+src_dir="$(pwd)"
```
You can use `$SRC_DIR`. Check all `conda-build`'s environment variables [here](http://conda.pydata.org/docs/building/environment-vars.html). They are very in handy sometimes.

ocefpaf commented on recipes/libradolan/build.sh:
```diff
@@ -0,0 +1,8 @@
+cd ../build
+cmake $src_dir -DCMAKE_INSTALL_PREFIX=$PREFIX
+make
+make install
```
Is there a `make check` or similar?

Can you also add a comment about the build? (It is unusual to have a two part build with `cmake` and then `make`.)

kmuehlbauer replied:
This is how it should be done according to the README:

- cmake
- make
- make install

cmake takes care of configuring everything, AFAIK, and creates the Makefile

I'm currently testing locally, because there is some problem with the CMakeLists.txt

ocefpaf commented on recipes/libradolan/meta.yaml:
```diff
@@ -0,0 +1,37 @@
+
+source:
+  git_url: https://github.com/meteo-ubonn/radolan.git
+  git_tag: 03e8107a51515230f603f91ec8a031f9b7035749
```
Can you change this to:

```yaml
{% set commit = "03e8107a51515230f603f91ec8a031f9b7035749" %}

fn: libradolan-{{ version }}.tar.gz
url: https://github.com/meteo-ubonn/radolan/archive/{{ commit }}.tar.gz
sha256: 7506ad48528dc02769685ae7a4fe82bd30e48aebb944ec45be69f58ffb0c7a81
```

ocefpaf commented:
> @ocefpaf Do you have any ideas, on how to address the architecture issue?

Sorry. I am out of my depth here. @jakirkham can probably help, but he is busy these days with his day job.

kmuehlbauer commented:
The search for `quadmath.h` brought me here: https://github.com/conda-forge/staged-recipes/pull/576#issuecomment-218307152

Looks like this solved the `quadmath` issue.

ccordoba12 commented:
The failure on Travis (Mac) is caused because of your use of gcc/libgcc, so please limit them to be only used on Linux :-)

kmuehlbauer commented:
Seems anaconda cloud has a problem right now, getting errors like this:

```
raise HTTPError(http_error_msg, response=self)
requests.exceptions.HTTPError: 404 Client Error: Not Found for url: https://conda.anaconda.org/conda-forge/
```

kmuehlbauer commented:
@ccordoba12 Thanks for the hint, I explicitely deactivated gcc and libgcc for osx.

Unfortunately I get this linking error complaining something about architecture (see below). I do not have access to OSX at the moment (@JuergenSimon, can you also have a look). Do I have to set something  more for osx?

```
[100%] Linking CXX executable radolan2netcdf

Undefined symbols for architecture x86_64:

  "boost::program_options::to_internal(std::string const&)", referenced from:

      std::vector<std::string, std::allocator<std::string> > boost::program_options::to_internal<std::string>(std::vector<std::string, std::allocator<std::string> > const&) in radolan2netcdf.cpp.o

  "boost::program_options::options_description::options_description(std::string const&, unsigned int, unsigned int)", referenced from:

      _main in radolan2netcdf.cpp.o

  "boost::program_options::invalid_option_value::invalid_option_value(std::string const&)", referenced from:

      void boost::program_options::validate<float, char>(boost::any&, std::vector<std::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::basic_string<char, std::char_traits<char>, std::allocator<char> > > > const&, float*, long) in radolan2netcdf.cpp.o

  "boost::program_options::error_with_option_name::error_with_option_name(std::string const&, std::string const&, std::string const&, int)", referenced from:

      std::basic_string<char, std::char_traits<char>, std::allocator<char> > const& boost::program_options::validators::get_single_string<char>(std::vector<std::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::basic_string<char, std::char_traits<char>, std::allocator<char> > > > const&, bool) in radolan2netcdf.cpp.o

  "boost::program_options::detail::cmdline::set_additional_parser(boost::function1<std::pair<std::string, std::string>, std::string const&>)", referenced from:

      boost::program_options::basic_command_line_parser<char>::extra_parser(boost::function1<std::pair<std::string, std::string>, std::string const&>) in radolan2netcdf.cpp.o

  "boost::program_options::detail::cmdline::cmdline(std::vector<std::string, std::allocator<std::string> > const&)", referenced from:

      boost::program_options::basic_command_line_parser<char>::basic_command_line_parser(int, char const* const*) in radolan2netcdf.cpp.o

  "boost::program_options::validate(boost::any&, std::vector<std::string, std::allocator<std::string> > const&, std::string*, int)", referenced from:

      boost::program_options::typed_value<std::string, char>::xparse(boost::any&, std::vector<std::string, std::allocator<std::string> > const&) const in radolan2netcdf.cpp.o

  "boost::program_options::operator<<(std::ostream&, boost::program_options::options_description const&)", referenced from:

      _main in radolan2netcdf.cpp.o

  "boost::program_options::abstract_variables_map::operator[](std::string const&) const", referenced from:

      _main in radolan2netcdf.cpp.o

  "boost::program_options::error_with_option_name::substitute_placeholders(std::string const&) const", referenced from:

      vtable for boost::exception_detail::clone_impl<boost::exception_detail::error_info_injector<boost::program_options::invalid_option_value> > in radolan2netcdf.cpp.o

      vtable for boost::exception_detail::error_info_injector<boost::program_options::invalid_option_value> in radolan2netcdf.cpp.o

      vtable for boost::program_options::invalid_option_value in radolan2netcdf.cpp.o

      vtable for boost::program_options::validation_error in radolan2netcdf.cpp.o

      vtable for boost::exception_detail::clone_impl<boost::exception_detail::error_info_injector<boost::program_options::validation_error> > in radolan2netcdf.cpp.o

      vtable for boost::exception_detail::error_info_injector<boost::program_options::validation_error> in radolan2netcdf.cpp.o

  "boost::program_options::value_semantic_codecvt_helper<char>::parse(boost::any&, std::vector<std::string, std::allocator<std::string> > const&, bool) const", referenced from:

      vtable for boost::program_options::typed_value<float, char> in radolan2netcdf.cpp.o

      vtable for boost::program_options::typed_value<std::string, char> in radolan2netcdf.cpp.o

ld: symbol(s) not found for architecture x86_64

clang: error: linker command failed with exit code 1 (use -v to see invocation)

make[2]: *** [radolan2netcdf] Error 1

make[1]: *** [CMakeFiles/radolan2netcdf.dir/all] Error 2

make: *** [all] Error 2

Command failed: /bin/bash -x -e /Users/travis/build/conda-forge/staged-recipes/recipes/libradolan/build.sh
``

kmuehlbauer commented:
OK, giving up for a while. Travis-CI didn't start at all, and CircleCi is experiencing outages from `aws`.

jakirkham commented:
Unfortunately, I can't be here long, but three things.

Why do we need `gcc`? The Linux container has `gcc` with C++11 support too. It's basically the same version. Also, `libgcc` can be dropped.

On Mac, adding something like [this]( https://github.com/jakirkham/staged-recipes/blob/f1f84a51b881555f342c4fcb07f5e0770fbe6a03/recipes/toolchain/activate.sh#L3-L14 ) to you `build.sh` should clean things up.

If you use C++11, adding `-std=c++11` to the `CXXFLAGS` for both *NIXes should help.

If it is still a mess, I should be able to take a closer look Tuesday.

kmuehlbauer commented:
> Why do we need gcc? The Linux container has gcc with C++11 support too. It's basically the same version. Also, libgcc can be dropped.

this was recommended here https://github.com/conda-forge/staged-recipes/pull/576#issuecomment-218307152, to overcome missing `quadmath.h` issue. 

If there is an other solution, I would gladly drop `gcc` at all.

To the other stuff: unfortunately I can't test this for osx, so I would have to iterate here. I'll try get things cleaned up a bit. If this won't work, I very much appreciate another look from @jakirkham next week.

jakirkham commented:
> ...to overcome missing `quadmath.h` issue.

Yeah, basically this means we are not installing `libquadmath` in the container, which is true. Right now using `gcc` on Linux only seems prudent. It's not clear how we want to address those cases. We are being pretty sparing with what we expect the end user to have, but when it comes to these basic `gcc` dependencies I'm not sure it is so prudent. Alternatively, if you don't need `float128` support here, you could probably disable it in the configure step. That would also get you out of having `libquadmath`.

kmuehlbauer commented:
@jakirkham Thanks for giving this insight.

> Alternatively, if you don't need float128 support here, you could probably disable it in the configure step. That would also get you out of having libquadmath.

I'm bound to use `cmake` here, but I'm neither an expert in `configure` nor in `cmake`. I'll try to find out how `float128` can be disabled.

kmuehlbauer commented:
I'm trying to investigate this locally on linux. When running `scripts/run_docker_build.sh` it works  out of the box, but it uses `/usr/bin/c++` instead the `/opt/rh/devtoolset-2/root/usr/bin/c++` provided by the docker-image. How can explicitely enable the docker-image compiler? @ocefpaf, @pelson Do you have any pointers, my search wasn't successful so far.

kmuehlbauer commented:
> How can explicitely enable the docker-image compiler?

Never mind, obviously I was using an obvious very ancient version of obvious-ci. Sorry for the noise.

ericdill commented on recipes/libradolan/build.sh:
```diff
@@ -0,0 +1,38 @@
+    export MACOSX_VERSION_MIN="10.7"
+    export MACOSX_DEPLOYMENT_TARGET="${MACOSX_VERSION_MIN}"
+    export CXXFLAGS="${CXXFLAGS} -mmacosx-version-min=${MACOSX_VERSION_MIN}"
+    export CXXFLAGS="${CXXFLAGS} -stdlib=libc++"
```
Try changing `-stdlib=libc++` to `-stdlib=libstdc++`

jakirkham replied:
Unfortunately, that won't work. `-stdlib=libstdc++` forces one to use `gcc`, which on Mac is 4.2.1.

ericdill commented on recipes/libradolan/build.sh:
```diff
@@ -0,0 +1,38 @@
+    export CXXFLAGS="${CXXFLAGS} -mmacosx-version-min=${MACOSX_VERSION_MIN}"
+    export CXXFLAGS="${CXXFLAGS} -stdlib=libc++"
+    export LDFLAGS="${LDFLAGS} -mmacosx-version-min=${MACOSX_VERSION_MIN}"
+    export LDFLAGS="${LDFLAGS} -stdlib=libc++ -lc++"
```
Try changing `-stdlib=libc++` to `-stdlib=libstdc++`. Not sure what `-lc++` is doing. Might not need it

jakirkham replied:
Same point with `libstdc++`.  `-lc++` does basically the exact same thing, but others have shown that this is sometimes also required. Though if it helps to drop this, please try.

kmuehlbauer commented:
@ericdill, @jakirkham thanks very much for review. I'm a bit out of options. Do we have some more things to try? Also @JuergenSimon should take a look, if he is available again.

jakirkham commented:
> Do we have some more things to try?

Yep, just one more. FWIW it worked in the docker container locally. Added it in a PR against your PR ( https://github.com/kmuehlbauer/staged-recipes/pull/1 ). If it still doesn't work, please let me know.

jakirkham commented:
Seems like Travis CI is hitting the same issue I was. See these [lines]( https://travis-ci.org/conda-forge/staged-recipes/jobs/130752897#L920-L921 ).

kmuehlbauer commented:
Yes, I'll have to look into the sources then. Is this something with C11 code incompatibility?

jakirkham commented:
> Yes, I'll have to look into the sources then. Is this something with C11 code incompatibility?

C11 or C++11 😜 Sorry we were talking about C++11 earlier so thought I should clarify.

Don't think this is that complicated. I have seen `clang` catch errors that `gcc` quietly swallows. Not sure what the assembly `gcc` produces looks like in those cases. Just suggesting this may be another case like this. From looking at the source pulled in by the recipe, I couldn't tell where `_Bool` was defined. Then again, this isn't my code. So I don't know it so well.

kmuehlbauer commented:
@jakirkham I tried to get things along and fixed the `_Bool`-problem locally. Now I get again those errors:

`ld: symbol(s) not found for architecture x86_64`

Just out of curiosity, because it uses netcdf-cxx4, it cannot be a problem of linking nectdf-cxx4 which was built with `gcc` (look here https://travis-ci.org/conda-forge/netcdf-cxx4-feedstock/builds/124251069)?

conda-forge-linter commented:
Hi! This is the friendly automated conda-forge-linting service.

I just wanted to let you know that I linted all conda-recipes in your PR (```recipes/libradolan```, ```recipes/netcdf-cxx4```) and found it was in an excellent condition.

kmuehlbauer commented:
@jakirkham This is a mess now, but I think I've found the problem (hopefully). The dependency `netcdf-cxx4` was built under MacOSX with some defaults which turned out to use `libstdc++` (with gcc, not clang). Therefore the linker errors.

I have added the netcdf-cxx4 recipe here and added your provided changes to check if this works. I have a PR ready at https://github.com/conda-forge/netcdf-cxx4-feedstock/pull/5. I'll clean up here if this is merged over there.

jakirkham commented:
Awesome. That sounds totally reasonable. Yes, I have seen this behavior with `gcc` before, which is one of many reasons why I want us to avoid that compiler if at all possible. As mentioned on that PR, one thing that should make this easier going forward is the `toolchain` package. As part of the cleanup here, we should switch to using the `toolchain` package here. Basically, just use this PR ( https://github.com/conda-forge/thrift-cpp-feedstock/pull/4 ) as a template. The only thing that may be different is you may need to add `CXXFLAGS="${CXXFLAGS} -std=c++11"` after sourcing here. We can totally do this after fixing `netcdf-cxx4` though.

jakirkham commented:
The error at the end (snippet shown below or highlighted in the [log]( https://travis-ci.org/conda-forge/staged-recipes/jobs/131099529#L5189-L5192 )) seems due to the `default` package coming in. We might need to do some pinning or something to handle this.

```
dyld: Library not loaded: @rpath/./libnetcdf.7.dylib
  Referenced from: /Users/travis/miniconda/envs/_test/bin/radolan2netcdf
  Reason: image not found
/Users/travis/miniconda/conda-bld/test-tmp_dir/run_test.sh: line 5:  7092 Trace/BPT trap: 5       radolan2netcdf --version
TESTS FAILED: libradolan-1.2.0-0
```

conda-forge-linter commented:
Hi! This is the friendly automated conda-forge-linting service.

I just wanted to let you know that I linted all conda-recipes in your PR (```recipes/libradolan```) and found it was in an excellent condition.

kmuehlbauer commented:
Crazy stuff, why is it pulling correctly for building:

 `netcdf-cxx4-4.2.1          |                2         338 KB  conda-forge`

but for running it gets:

`netcdf-cxx4-4.2.1          |                0         326 KB  conda-forge`

The idea with pinning seems reasonable, we could add libnetcdf or hdf5 as dependency (will be needed anyway) and pin this instead.

I'll try this next.

jakirkham commented:
> Crazy stuff, why is it pulling correctly for building...

That is a question that I have asked many times before. I'm really suspicious that there is something broken about how the testing portion of `conda-build` handles resolutions.

jakirkham commented:
For some interesting context on `boost`, `gcc` vs. `clang`, and other fun Mac things, check out this [comment]( http://stackoverflow.com/a/25424412 ) on SO and related links.

kmuehlbauer commented:
Finally, this is working.

I had to pin `hdf5` to `1.8.15*` to get the correct `netcdf-cxx4` pulled for the `run`. Will create issue in the feedstock, when this gets merged.

ping @jakirkham, @ocefpaf Could you please have a final look, and merge if LGTY.

ocefpaf commented on recipes/libradolan/build.sh:
```diff
@@ -0,0 +1,38 @@
+    export LINKFLAGS="${LDFLAGS}"
+else
+    echo "This system is unsupported by the toolchain."
+    exit 1
```
:+1:

ocefpaf commented:
Thanks @kmuehlbauer! This looks good to me.
